### DISCUSSION TITLE
How to unit test sequelize ?

### DISCUSSION THREAD

**Discussion by TannerS** [sequelize/sequelize#16037] [Help & Questions]
Hello we have a NPM module that wraps around sequelize. it basically defines our models and connections and etc for our services in muliple microservices. all internal, nothing public. But as an npm module that talks to a database, any advice on how to unit test this module using sequelize ?

**Comment by MarioRomanDono** [comment]
You should try to separate your code in two parts: your own code and the infrastructure code (meaning the Sequelize code). That way, you could test your module like this:

- Unit test your code, mocking the infrastructure. That way, you can simulate the responses and errors, without depending on the real database.
- Create integration tests on the infrastructure code, checking that data is being stored, retrieved and updated correctly. Here you don't want to mock the database: you want to use Sequelize and the real database (using a local database or Docker).

You can read more about this here: https://martinfowler.com/articles/practical-test-pyramid.html
